design/ceph/ceph-telemetry.md

```diff
@@ -0,0 +1,101 @@
+- Users must opt in to telemetry
+- Users must also opt into the new (delta) telemetry items added between Ceph versions
+- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
+  usage of the mon database low
+
+## Proposed metrics for Rook to collect
+- `rook-version` - Rook version.
+  - Done in https://github.com/rook/rook/pull/10161
+- `rook-kubernetes-version` - Kubernetes version.
+  - This is similar to the already collected os/kernel versions.
+- `rook-num-kubernetes-nodes` - Number of Kubernetes nodes.
+- `rook-ceph-csi-version` - Ceph CSI version.
+- Num CSI plugin pods running - This gives idea of how many k8s nodes are in the k8s cluster that
+  are allowed to use Ceph storage.
+  - It is easiest count plugin pods separately since each DaemonSet is a different label. Plus, it
+    can show if some CSI drivers are ever disabled.
+  - `rook-num-csi-rbd-plugin-pods`
+  - `rook-num-csi-cephfs-plugin-pods`
+  - `rook-num-csi-nfs-plugin-pods`
+- `rook-num-nodes` - Number of nodes on which the cluster has running pods.
+  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
+    count for Rook installs in portable environments
+  - We can get this by counting the number of crash collector pods; however, if any users disable
+    the crash collector, we should report an unknown value, possibly `unknown`.
+- `rook-num-storage-classes` - Number of storage classes related to the current Ceph cluster. We can
+  also break this up by specific kinds of storage classes, but the overall may be useful if new
+  classes are added in the future.
+  - `rook-num-rbd-storage-classes`
+  - `rook-num-cephfs-storage-classes`
+  - `rook-num-nfs-storage-classes`
+  - `rook-num-bucket-storage-classes`
+- `rook-num-storage-class-device-sets` - Number of storage class device sets.
+- `rook-num-portable-storage-class-device-sets` - .
+- `rook-mons-are-pvc` - true/false whether mons are on PVC.
+- `rook-network-provider` - The network provider used for the cluster. e.g., default, host, multus.
+- `rook-external-mode` - true/false if the cluster is in external mode.
+
+## Information Rook is interested in that is already collected by Ceph telemetry
+- RBD pools (name is stripped), some config info, and the number of them.
+- Ceph Filesystems and MDS info.
+- RGW count, zonegroups, zones.
+- RBD mirroring info.
+
+## Metrics that Rook would like to see added to Ceph's telemetry
+- Number of NFS clusters
+  - Ceph can count NFS clusters by listing the namespaces in the `.nfs` pool and report those
+    namespaces that have `conf-nfs.<namespace>` objects present
+- Filesystem mirroring info
+- RGW bucket notification usage (Yuval may have good detail on this point)
+
+## Proposed integration strategy
+- Rook will first merge [`rook-version`](https://github.com/rook/rook/pull/10161).
+- Rook, with input from Ceph telemetry, will approve a version of this design doc and treat all
```

Comment:
Instead of locking on this design doc, I'm thinking the telemetry const names could be the source of coordination. Design docs are just never perfectly updated with the code.

Reply:
In most cases, I would agree. But since this is a contract we will maintain with the Ceph team, I think we should treat this doc as gospel in this case. Once Ceph implements these, we should never change them.

Reply:
That would mean we could only add metrics to this design doc if they are implemented in a given batch, right? But capturing future possible metrics seems like a good and necessary thing. 

How strict of a contract is really necessary? If the values are just all key-value pairs, and the telemetry will only gather the keys we've told them about, it seems sufficient to look at the consts in the code. Even if we remove metrics in the future, I expect it won't break the telemetry, it would just not be gathered anymore.